packages/next/src/server/base-server.ts

```diff
@@ -638,14 +638,6 @@ export default abstract class Server<ServerOptions extends Options = Options> {
     }
 
     if (params.path[0] !== this.buildId) {
-      // Ignore if its a middleware request when we aren't on edge.
-      if (
-        process.env.NEXT_RUNTIME !== 'edge' &&
-        req.headers['x-middleware-invoke']
-      ) {
-        return false
```

Comment:
Removing this breaks the functionality that allows middleware to rewrite `_next/data` paths for a different project/deployment. 

Instead of changing this, we should ensure we 404 properly after middleware is invoked, we didn't rewrite/redirect, and the buildId doesn't match at that point.

Reply:
First, I am sorry about the failing test cases. I tried to get `pnpm build` and `pnpm test-start` working in 3 different envs (macOS (with volta installed), the repo's devcontainer, and a ubuntu VM) with `pnpm build` complaining about missing dependencies after `pnpm i`).

> Instead of changing this, we should ensure we 404 properly after middleware is invoked, we didn't rewrite/redirect, and the buildId doesn't match at that point.

But the case of a successful rewrite is the culprit here. next.js needs to properly 404 after middleware is invoked, the buildId is different and a rewrite happened.

Clicking Link after a deployment with a different buildId will do the following:
Request to `_next/data/....json` is made with:
```
'x-middleware-prefetch': '1'
'x-nextjs-data': '1'
```

Middleware will rewrite that and will add
```
'x-invoke-path': '',
'x-invoke-query': '',
'x-invoke-output': '',
'x-middleware-invoke': '1'
```
before `normalizeAndAttachMetadata` and thus `handleNextDataRequest` is called.

Based on these headers the issue is in the branch where `handleCatchallRenderRequest` is called.
I am not sure if this is the only branch that is affected, though, as I am not very familiar with "middleware to rewrite _next/data paths for a different project/deployment."

What about letting `normalizeAndAttachMetadata` return `[complete, buildIdMismatch]`, and before `handleCatchallRenderRequest` check if `buildIdMismatch` is true and render 404 in that case before executing `handleCatchallRenderRequest`?


I can provide a reproducible case if you want.